### ISSUE TITLE
[TBC] Creatures calling for help (fleeing) is ignoring Line of Sight (Crossing walls)

### ISSUE THREAD

**Issue by SilvioDoMine** [cmangos/issues#1720]
**Current behavior**: 
If a creature starts flee calling for help (and running towards them) he will ignore Line of Sight and will cross walls.

**Expected behavior**:
He shouldn't ignore Line of Sight and do a proper way to flee.

**Steps to reproduce**:

1. take a char level 15 (.level 14)
2. type .go xyz -9847.349609 1053.160034 33.589199
3. engage combat with the creature inside the house
4. Attack him until he loses hp and starts calling for help
5. He will run towards the wall and cross it

**Client version**: 
2.4.3 (8606)

**Commit hash**: 
 8acf1bbc83b7239a24cc7872cdab31d889e60634

**Database version**: 
Latest

**Operating system**: 
Linux Debian 9

[//]: # (If this is a crash report, include the crashlog from a debug build with https://gist.github.com/)

**Comment by bdebaere** [comment]
Same on classic.

**Comment by Grz3s** [comment]
same mehanic happens in all 3 cores

**Comment by yatzi888** [comment]
I see to also experience this with feared mobs. If that also counts as "fleeing" TBC 

Client version: 2.4.3

Commit hash: c559b36d4e99345002f59987161bb87591bc30d4

Database version: 8de43372a43d824a3f9e5688d1f6609ab71cb775

**Comment by LevelOneX** [comment]
Is this VMaps or object collision code related ?

**Comment by Warlockbugs** [comment]
This is a problem with maps. Nothing close to this was occurring a few months ago.

**Comment by killerwife** [comment]
Calling of mobs is done inside "MoveInLineOfSight", and as a result it uses LOS calculations. There must be something else that is problematic.

**Comment by Warlockbugs** [comment]
Its not calling for help that is a problem at the moment, the pathfinder routes units through wmos. I cant recall any recent changes to pathfinder code that could affect it, but we had format update for data.

**Comment by Ve1vet** [comment]
@killerwife can be this https://github.com/cmangos/mangos-tbc/commit/6430d90ff668b44db25f456548d107312f3cfbae ?

**Comment by killerwife** [comment]
No, bcs MoveInLineOfSight is only called when ur in actual LOS. So the func would not be entered at all.

**Comment by MagPurify** [comment]
I am tracking this problem for a long time, and I really want to help. I have checked other repos to see how they have done object collision and PathFinder, but everything that I see, usually needs to import so much code that I just get afraid to touch anything. Though, I heard LH & Nolst lag were related to PathFinder, so since this is such a sensitive code to play, not sure what to do.

**Comment by AnonXS** [comment]
Add command ".gps" data else this really leads nowhere

**Comment by MagPurify** [comment]
is it better to re-upload the above scenarios with command gps into the chat ? Sorry I didn't know that command.

**Comment by AnonXS** [comment]
just write the .go xyz COORDs in the chat beow the screenshots, so one can lookup the locations more quickly.

**Comment by killerwife** [comment]
@MagPurify and did you do those screenshots with my change to fleeing movement generator?

**Comment by MagPurify** [comment]
I will do that @AnonXS , thanks for the tip
Yes @killerwife, I was testing your fix, I see it a little harder for creatures to cross now, but unfortunately it is still doing that.

**Comment by killerwife** [comment]
Try using MovePositionToFirstCollision inside Fleeing Movement Generator and see if it fixes it.

**Comment by MagPurify** [comment]
Should I replace the collision `owner.GetMap()->GetHitPosition` that is has there with the `MovePositionToFirstCollision` ? or do that before updating the position inside the IF block ?

**Comment by killerwife** [comment]
both the updateallowedpositionz and the gethitposition should be replaced.

**Comment by MagPurify** [comment]
<img width="1280" alt="newchangefleeing" src="https://user-images.githubusercontent.com/46841019/51428759-a66f7700-1bee-11e9-9943-108fc1d36227.png">

```
    float angle = frand(0, 2 * M_PI_F);
    float dist;

    if (dist_from_caster < MIN_QUIET_DISTANCE)
        dist = frand(0.4f, 1.3f) * (MIN_QUIET_DISTANCE - dist_from_caster);
    else if (dist_from_caster > MAX_QUIET_DISTANCE)
        dist = frand(0.4f, 1.0f) * (MAX_QUIET_DISTANCE - MIN_QUIET_DISTANCE);
    else    // we are inside quiet range
        dist = frand(0.6f, 1.2f) * (MAX_QUIET_DISTANCE - MIN_QUIET_DISTANCE);

	WorldLocation worldLocation;
	owner.GetFirstCollisionPosition(worldLocation, dist, angle);
	x = worldLocation.coord_x;
	y = worldLocation.coord_y;
	z = worldLocation.coord_z;

    return true;
}
```

I did this change above, I believe it had the same effect. Crossing walls on X or Y avoids well. The problem is the Z, when something is above or below.

**Comment by MagPurify** [comment]
Tested on the cave, it worked well, sometimes they climb the walls, but so far it haven't crossed.

**Comment by MagPurify** [comment]
I was able to reproduce that @LevelOneX. The Z problem might be related to the MoveChase from NPC as well. The attached screenshot is a example of when I was running with a NPC behind me, he started to flew to the above path. Is this a different problem @killerwife ?
<img width="1280" alt="movechase" src="https://user-images.githubusercontent.com/46841019/51429673-12a2a880-1bf8-11e9-8880-7cc6f9248050.png">

**Comment by killerwife** [comment]
That one is bad mmap extraction I'm afraid. Look at it through RecastDemo from the repo. ".mmap loc" will get u info. And then you need to load up vmaps and mmaps to see how it looks over there.

**Comment by MagPurify** [comment]
Thanks for everything @killerwife, you rock. One more question, since I don't want to abuse it. What's the best way to fix blink ? For example, in UC It's easy to fall into the doom. In houses, tents, or cave, when I blink sometimes I get above or below the place.

**Comment by MagPurify** [comment]
This is also a sample of what is happening with the above logic as well.
I believe should fall into the same category. 

https://drive.google.com/file/d/1ohSJ4Wjh3OJ__ReW_XFlGAo-JJNZ6GfN/view?usp=sharing

**Comment by AnonXS** [comment]
I can not reprodue the above mentioned npc flee or player fear through buildings/walls/wmo/m2 anymore.

Fact is that some npcs dont do random flee at low hp but infact do a targeted flee at low hp which we've to launch through dbscript currently.

**Comment by AnonXS** [comment]
Can anyone provide good reproduceable situations for "Flee through Wall" and "Fear through Wall"?
